x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/CreateEnrollmentToken.java

```diff
@@ -0,0 +1,284 @@
+        Map<?, ?> http = (Map<?, ?>) nodeInfo.get("http");
+        return (ArrayList<String>) http.get("bound_address");
+    }
+
+    static String getVersion(Map<?, ?> nodesInfo) {
+        nodesInfo = (Map<?, ?>) nodesInfo.get("nodes");
+        Map<?, ?> nodeInfo = (Map<?, ?>) nodesInfo.values().iterator().next();
+        return nodeInfo.get("version").toString();
+    }
+
+    protected String getApiKey(String user, SecureString password) throws Exception {
+        final CheckedSupplier<String, Exception> createApiKeyRequestBodySupplier = () -> {
+            XContentBuilder xContentBuilder = JsonXContent.contentBuilder();
+            xContentBuilder.startObject()
+                .field("name", "enrollment_token_API_key_" + UUIDs.base64UUID())
+                .field("expiration", ENROLL_API_KEY_EXPIRATION)
+                .startObject("role_descriptors")
+                .startObject("create_enrollment_token")
+                .field("cluster", "[cluster:admin/xpack/security/enrollment/enroll/node]")
+                .endObject()
+                .endObject()
+                .endObject();
+            return Strings.toString(xContentBuilder);
+        };
+
+        final URL url = new URL(client.getDefaultURL());
+        final URL Url = createAPIKeyURL(url);
+        final HttpResponse httpResponseApiKey = client.execute("POST", Url, user, password,
+            createApiKeyRequestBodySupplier, is -> responseBuilder(is));
+        final int httpCode = httpResponseApiKey.getHttpStatus();
+
+        if (httpCode != HttpURLConnection.HTTP_OK) {
+            throw new IllegalStateException("Unexpected response code [" + httpCode + "] from calling POST "
+                + Url);
+        }
+
+        final String apiKey = Objects.toString(httpResponseApiKey.getResponseBody().get("api_key"), "");
+        if (Strings.isNullOrEmpty(apiKey)) {
+            throw new IllegalStateException("Could not create an api key.");
+        }
+        return apiKey;
+    }
+
+    protected Tuple<List<String>, String> getAddressesAndVersion(String user, SecureString password) throws Exception {
+        final CheckedSupplier<String, Exception> getHttpInfoRequestBodySupplier = () -> {
+            XContentBuilder xContentBuilder = JsonXContent.contentBuilder();
+            xContentBuilder.startObject()
+                .endObject();
+            return Strings.toString(xContentBuilder);
+        };
+
+        final URL url = new URL(client.getDefaultURL());
+        final HttpResponse httpResponseHttp = client.execute("GET", getHttpInfoURL(url), user, password,
+            getHttpInfoRequestBodySupplier, is -> responseBuilder(is));
+        final int httpCode = httpResponseHttp.getHttpStatus();
+
+        if (httpCode != HttpURLConnection.HTTP_OK) {
+            throw new IllegalStateException("Unexpected response code [" + httpCode + "] from calling GET " + getHttpInfoURL(url));
```

Comment:
nit: I've previously made a comment around calling getHttoInfoURL twice. Maybe a slightly more informative message like "Failed to call Node Info API , HTTP status code was such and such " ? The fact this is an (HTTP) GET is not very useful to the user or admin. Same for the error message for api keys